Sources/Core/Shared/GroupOperation.swift

```diff
@@ -339,3 +448,117 @@ public struct WillAddChildObserver: GroupOperationWillAddChildObserver {
+                    // by checking for child operations that are not finished.
+
+                    let nonFinishedOperations = parent.operations.filter({ !$0.finished })
+                    if !nonFinishedOperations.isEmpty {
+
+                        // Child operations were added after this CanFinishOperation became
+                        // ready, but before it executed or before the lock could be acquired.
+                        //
+                        // The GroupOperation should wait for these child operations to finish
+                        // before finishing. Add the oustanding child operations as
+                        // dependencies to a new CanFinishOperation, and add that as the
+                        // GroupOperation's new CanFinishOperation.
+
+                        let newCanFinishOp = GroupOperation.CanFinishOperation(parentGroupOperation: parent)
+
+                        nonFinishedOperations.forEach { op in
+                            newCanFinishOp.addDependency(op)
+                        }
+
+                        parent.canFinishOperation = newCanFinishOp
+
+                        parent.queue._addCanFinishOperation(newCanFinishOp, finishingOperation: parent.finishingOperation)
+                    }
+                    else {
+                        // There are no additional operations to handle.
+                        // Ensure that no new operations can be added.
+                        parent.isGroupFinishing = true
+                    }
+                    return false
+                }
+
+                guard !isWaiting else { return }
+            }
+
+            willChangeValueForKey("isFinished")
+            _finished = true
+            didChangeValueForKey("isFinished")
+        }
+        private func NSQualityOfServiceToDispatchQOS(qualityOfService: NSQualityOfService) -> qos_class_t {
+            let mapping = [NSQualityOfService.Background: QOS_CLASS_BACKGROUND,
+                           NSQualityOfService.Default: QOS_CLASS_DEFAULT,
+                           NSQualityOfService.UserInitiated: QOS_CLASS_USER_INITIATED,
+                           NSQualityOfService.UserInteractive: QOS_CLASS_USER_INTERACTIVE,
+                           NSQualityOfService.Utility: QOS_CLASS_UTILITY]
+            return mapping[qualityOfService] ?? QOS_CLASS_DEFAULT
+        }
+        override var executing: Bool {
+            return false
+        }
+        override var finished: Bool {
+            return _finished
+        }
+    }
+}
+
+private extension OperationQueue {
+    private func _addCanFinishOperation(canFinishOperation: GroupOperation.CanFinishOperation, finishingOperation: NSOperation) {
+        // Do not add observers (not needed - CanFinishOperation is an implementation detail of GroupOperation)
+        // Do not add conditions (CanFinishOperation has none)
+        finishingOperation.addDependency(canFinishOperation)
```

Comment:
Hmm, maybe it would be better for the `GroupOperation` or `CanFinishOperation` to manage it's dependencies, rather than the queue?